### DISCUSSION TITLE
How to Auotomatically restart fastapi application programmatically without manually stopping and start

### DISCUSSION THREAD

**Discussion by kvalli** [fastapi/fastapi#12191] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import fastapi

app = FastAPI()

@app.post("/Shutdown_ApiGw")
def shutdown():
    os.kill(os.getpid(), signal.SIGTERM)
    return fastapi.Response(status_code=200, content='Server shutting down...')

def restart():
    ??

uvicorn.run(app, host='localhost', port=8000)
```

### Description

I have a fast-api aplication which is running in docker.
There is a unity game application sending requests to 3  end points  in fastapi application.
Each end point is receiving requests every 0.1 second roughly for a period ranging from 3-5 minutes.
the main issue is the memory usage that is piling with this set up.
for one end point has to collect data for 5 mins and give a response.
fast api application is started by lanuching docker up and unity game also started.
Unity game completes one game and starts the new game but after one game the fast api application remains unresponsive and memory usage is piling up.
when unity game starts new game ,i want to restart fastapi application without manual intervention.How to achieve this?
I also have another query
How to clean up global and local variables in fast api application either after each request or when it is about to restart
Is it possible to have restart end point just like shutdown?

### Operating System

Windows

### FastAPI Version

0.110.1

### Pydantic Version

2.6.4

### Python Version

3.12.2

**Comment by nmatter1** [comment]
Start the server in a separate thread or use multiprocessing

**Comment by Irfan-Ahmad-byte** [comment, marked as answer]
If i'm right, with each new game there will be a new session. I this case here is what I think could be of use:

1. You can add a restart policy to your Dockerfile like:
```
services:
  fastapi_app:
    image: your_fastapi_image
    restart: always
    ports:
      - "8000:8000"

```

2. For local variables, the modules take care of them. For global variables, you can create a function to reset them all. You can create a separate script and run that script with each Dockerfile each time it starts/restarts.

3. a restart endpoint could be like this
```
@app.post("/restart")
async def restart_app():
    os.execv(sys.executable, ['python'] + sys.argv)
    return {"message": "Restarting"}

```
This replaces the current process in Python replacing with the new one, which is the same process.
